### DISCUSSION TITLE
I'm new to neovim and I don't know how to configure a formatter

### DISCUSSION THREAD

**Discussion by JasterV** [LazyVim/LazyVim#611] [Q&A] [answered]
Hello there :)

I'm new to neovim and did never configure a formatter before.
I want to customize the `shfmt` formatter to indent using only 2 spaces instead of 4.
I know that `shfmt` has an option to set that by passing `-i 2` as arguments.
I have no idea how to configure this so I would appreciate some help, I'm sure it should be straight forward

**Comment by folke** [comment]
I assume you're using null-ls with that formatter? Best to check the null-ls docs on how to configure formatter settings

**Comment by dpetka2001** [comment]
Like @folke said you should take a look at null-ls [wiki](https://github.com/jose-elias-alvarez/null-ls.nvim/wiki/Source-specific-Configuration) which has configurations about specific sources and customize accordingly your spec.

**Comment by JasterV** [comment]
Hi! I just installed LazyVim and I don't see any `null-ls` configuration on the default `example.lua`

I installed `shfmt` using `Mason` and thats it, I don't know much more

Does LazyVim use `null-ls` behind the scenes?

**Reply by folke** [reply]
Please read the docs https://www.lazyvim.org/ I created them for a reason

**Reply by JasterV** [reply]
Hey man... I did :) Don't be so rude I'm trying to get into nvim using LazyVim and there are many things that I don't understand

I see there is a section where it talks about `null-ls` but it does not specify if that is the internal default configuration or if I should copy-paste that into my own configuration

**Reply by JasterV** [reply]
I mean, https://www.lazyvim.org/plugins/lsp#null-lsnvim What am I supposed to do with this? I don't want to mess the default LazyVim configuration because it works amazing (I really like LazyVim, its helping me a lot), and this looks like I would override the default `null-ls` config if there is any

**Reply by dpetka2001** [reply]
I provided a link in my previous post with example configurations about specific sources. Seems like you need `extra-args` to do what you want for shfmt. Did you try that to see if it works?

**Reply by folke** [reply]
You don't need to copy-paste that. Those are the defaults indeed.

All the plugins under the plugins section are included in LazyVim, except for the ones under **extras**.

You should still read the docs. Both the ones for LazyVim and lazy.nvim.

There's some specific examples on configuring null-ls. How to configure a specific null-ls formatter is something you'll be abe to find in the null-ls docs.

**Reply by dpetka2001** [reply, marked as answer]
You need to create your own spec in `~/.config/nvim/lua/plugins` for extending null-ls functionality. You could experiment with something like this in a file created in the directory mentioned:
```lua
return {
  -- Setup null-ls with `prettierd`
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        nls.builtins.formatting.prettierd.with({
          filetypes = { "html", "css", "json", "jsonc", "yaml", "markdown" },
        }),
      })
    end,
  },
}
```
Just change `prettierd` to `shfmt` and use the `extra-args` instead of `filetypes` as mentioned in null-ls docs provided in my previous post. It should work this way although I haven't tested it.

**Reply by JasterV** [reply]
Perfect thank you so much to both of you, this is exactly what I was looking for!
🚀

**Reply by dpetka2001** [reply]
You should also read again both Lazyvim and lazy.nvim docs to start getting better understanding how things work and come together. I've read them more than just one time to get a better understanding myself and there are already many examples of what you want to do.

**Reply by JasterV** [reply]
I will definitely do, thank you 👍🏻

**Comment by emilecantin** [comment]
Hi, I will hijack this discussion as my issue is similar. I have a `.prettierrc` file in my project and I want neovim to format on save using the rules defined in it.

Looking at the [documentation](https://www.lazyvim.org/plugins/extras/formatting.prettier), looks like I can just add this extra to my LazyVim imports:
```
{ import = "lazyvim.plugins.extras.formatting.prettier" },
```

Unfortunately, that doesn't look like it's enough. The doc then goes on to talk about Mason.nvim (I don't know what it is, and its readme tells me it's a package manager, which I don't understand why I'd need if I already have LazyVim) with absolutely no context, and then null-ls (I kinda know about this one, but I never managed to configure it successfully). I'm not sure if these are 2 options I can choose, or if I have to have both.

Can somebody explain it a bit more? I don't care about a specific tool (null-ls vs prettierd vs whatever else), but I'd like the simplest way to set this up.

**Reply by JasterV** [reply]
Hi @emilecantin 
Unfortunately I'm quite new to neovim so I don't have the solution to this formatting thing, but I can definitely explain the need of [Mason](https://github.com/williamboman/mason.nvim)

Mason is a *Package* manager, while `lazy.nvim` is a *Plugin* manager. `LazyVim` is basically a Neovim configuration (A bunch of 
configuration files that make Neovim feel like a spaceship) that relies on `lazy.nvim` as the default plugin manager.

Mason helps you installing packages/tools in your system in an automatic way so you don't have to worry about doing it manually. These are tools like: Language servers, formatters, linters and more. 
Without Mason you would have to install these things manually in your system.

For example lets say you want to start programming in Rust. You will need the Rust language server so you can then configure Neovim to use it and get advantage of it. You can just use Mason to install it 🚀 

Then it comes the plugin manager. `lazy.nvim` allows you to install and manage plugins on a very user friendly way (Compared to other plugin managers in my opinion). For example, in order to make use of the language servers you have installed on your system, you would want to configure Neovim so it starts one or another depending on the extension of the file you are working on.
For this we have plugins like this [one](https://github.com/neovim/nvim-lspconfig) which comes installed and preconfigured in LazyVim by default.

Again, I'm kinda new into the Neovim ecosystem and this explanation might not be 100% accurate, but this is the way I understand it
